lib/chef/resource.rb

```diff
@@ -877,142 +877,144 @@ def after_created
-          @resource_name = value.to_sym
-          provides self.resource_name
-        end
-        # Backcompat: set resource name for classes in Chef::Resource automatically
-        if !@resource_name && self.name
-          chef, resource, class_name, *extra = self.name.split('::')
-          if chef == 'Chef' && resource == 'Resource' && extra.size == 0
-            @resource_name = convert_to_snake_case(self.name.split('::')[-1])
+    #
+    # The display name of this resource type, for printing purposes.
+    #
+    # This also automatically calls "provides" to provide DSL with the given
+    # name.
+    #
+    # resource_name defaults to your class name.
+    #
+    # Call `resource_name nil` to remove the resource name (and any
+    # corresponding DSL).
+    #
+    # @param value [Symbol] The desired name of this resource type (e.g.
+    #   `execute`), or `nil` if this class is abstract and has no resource_name.
+    #
+    # @return [Symbol] The name of this resource type (e.g. `:execute`).
+    #
+    def self.resource_name(name=NULL_ARG)
+      # Setter
+      if name != NULL_ARG
+        remove_canonical_dsl
+
+        # Set the resource_name and call provides
+        if name
+          name = name.to_sym
+          # If our class is not already providing this name, provide it.
+          if !Chef::ResourceResolver.list(name).include?(self)
+            provides name, canonical: true
           end
+          @resource_name = name
+        else
+          @resource_name = nil
         end
-        @resource_name
-      end
-      alias :resource_name= :resource_name
-
-      #
-      # Use the class name as the resource name.
-      #
-      # Munges the last part of the class name from camel case to snake case,
-      # and sets the resource_name to that:
-      #
-      # A::B::BlahDBlah -> blah_d_blah
-      #
-      def use_automatic_resource_name
-        automatic_name = convert_to_snake_case(self.name.split('::')[-1])
-        resource_name automatic_name
-      end
-
-      #
-      # The module where Chef should look for providers for this resource.
-      # The provider for `MyResource` will be looked up using
-      # `provider_base::MyResource`.  Defaults to `Chef::Provider`.
-      #
-      # @param arg [Module] The module containing providers for this resource
-      # @return [Module] The module containing providers for this resource
-      #
-      # @example
-      #   class MyResource < Chef::Resource
-      #     provider_base Chef::Provider::Deploy
-      #     # ...other stuff
-      #   end
-      #
-      # @deprecated Use `provides` on the provider, or `provider` on the resource, instead.
-      #
-      def provider_base(arg=nil)
-        if arg
-          Chef::Log.deprecation("Resource.provider_base is deprecated and will be removed in Chef 13. Use provides on the provider, or provider on the resource, instead.")
+      else
+        # set resource_name automatically if it's not set
```

Comment:
I was going to ask if this is the place where the automagic happens, but it seems you've left a comment to tell me that